books/RayTracingTheRestOfYourLife.html

```diff
@@ -1680,67 +1679,81 @@
-Instead of having our scattering PDF perfectly match the Lambertian distribution  -- again --
-$\cos(\theta_o)$, we'll just use a uniform pdf about the hemisphere, $1/2\pi$. This will still
-converge on the correct answer, all we've done is change the PDF, but since the PDF is now less of a
-perfect match for the real distribution, it will take longer to converge:
+Now, just for the experience, let's try using a different sampling PDF. We'll continue to have our
+reflected rays weighted by Lambertian, so $\cos(\theta_o)$, and we'll keep the scattering PDF as is,
+but we'll use a different PDF in the denominator. We will sample using a uniform PDF about the
+hemisphere, so we'll set the denominator to $1/2\pi$. This will still converge on the correct
+answer, as all we've done is change the PDF, but since the PDF is now less of a perfect match for
+the real distribution, it will take longer to converge. Which, for the same number of samples means
+a noisier image:
 
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
-    class lambertian : public material {
-      public:
+    class scene {
         ...
-        bool scatter(
-            const ray& r_in, const hit_record& rec, color& alb, ray& scattered, double& pdf
-        ) const override {
-            auto scatter_direction = rec.normal + random_unit_vector();
+      private:
+        color ray_color(const ray& r, int depth) {
+            hit_record rec;
 
-            // Catch degenerate scatter direction
-            if (scatter_direction.near_zero())
-                scatter_direction = rec.normal;
+            // If we've exceeded the ray bounce limit, no more light is gathered.
+            if (depth <= 0)
+                return color(0,0,0);
 
-            scattered = ray(rec.p, unit_vector(scatter_direction), r_in.time());
-            alb = albedo->value(rec.u, rec.v, rec.p);
-    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
-            pdf = 0.5 / pi;
-    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
-            return true;
-        }
+            // If the ray hits nothing, return the background color.
+            if (!world.hit(r, interval(0.001, infinity), rec))
+                return background;
 
-        double scattering_pdf(const ray& r_in, const hit_record& rec, const ray& scattered) const {
+            ray scattered;
+            color attenuation;
+            color emitted = rec.mat->emitted(rec.u, rec.v, rec.p);
+
+            if (!rec.mat->scatter(r, rec, attenuation, scattered))
+                return emitted;
+
+            double scattering_pdf = rec.mat->scattering_pdf(r, rec, scattered);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
-            return 0.5 / pi;
+            double pdf = 1 / (2*pi);
```

Comment:
I think a note should be made somewhere that the pdf and scattering pdf are currently cancelling each other out. By multiplying by `0.5 / pi`, and then dividing by `0.5 / pi` in the same expression.

![image](https://user-images.githubusercontent.com/1538891/148639152-2ac03f83-d6a1-4254-b894-cbc662eb5631.png)

Reply:
?

I think you're looking over the book before this CL.
I explicitly call out that they cancel each other in the latest change

Reply:
Awesome. It's great that you've called this observation out. to make it clear

Reply:
Feel free to make as Resolved (I don't seem to have that option)

Reply:
Cool. Thanks for the feedback!